docs/vm-configuration.md

```diff
@@ -266,3 +266,227 @@ the generation of multiple libvirt XML documents, the domain XML and associated
+
+Disks and network interfaces are good examples where KubeVirt provides
+specializations. See the mapping decisions for them below to get an impression
+on how clear extension points can be added. In general allowing to choose
+between different optional source or target structs is the preferred mechanism,
+to cleanly separate between KubeVirt extensions and libvirt source or target
+types.
+
+For device types where no KubeVirt specific extensions are expected, creating
+sub-structs is not required. Even differentiating based on `type` fields is OK
+there. Examples are `smartcard`, `input` and `controller`.
+
+### Use optional structs to represent different sources or targets
+
+By having an inlinded wrapper-struct, containing optional structs corresponding
+to different types, we get a reasonable usable API and can maintain explicit
+mapping withing the code.
+
+Given a `Disk` which can have different `DiskSource`s
+
+```golang
+type Disk struct {
+    Name string `json:"name"`
+    DiskSource `json:"diskSource,inline"`
+}
+
+type DiskSource struct {
+   ISCSI *ISCSIDiskSource `json"iscsi,omitempty"`
+   NoCloud *NoCloudDiskSource `json"noCloud,omitempty"`
+}
+```
+
+it will map to
+
+```yaml
+disk:
+  name: mydisk
+  diskSource:
+    iscsi:
+      blub: bla
+disk:
+  name: mydisk
+  diskSource:
+    noCloud:
+      blub: bla
+```
+
+## Mapping decisions for specific device types
+
+### Disks
+
+`disk`, `lun`, `floppy` and `cdrom` aret top level entries in the `devices`
+array. Based on them, the `*.target` section can therefore be specialized based
+on the top level struct. Further, the `*.source` section can be specialized and
+can contain structs for traditional sources like `iscsi` or KubeVirt specific
+structs like `noCloud`:
+
+```yaml
+devices:
+- disk:
+    target:
+      readOnly: true
+    source:
+      noCloud:
+        secretRef:
+          name: mysecret
+- cdrom:
+    target:
+      readOnly: true
+      tray: "open"
+    source:
+      image:
+        name:"mydisk:latest"
+- floppy:
+    target:
+      readOnly: true
+      tray: "open"
+    source:
+      image:
+        name:"mydisk:latest"
+- lun:
+    target:
+      readOnly: false
+    source:
+      iscsi:
+        targetPortal: 10.0.2.15:3260
+        portals: ['10.0.2.16:3260', '10.0.2.17:3260']
+        iqn: iqn.2001-04.com.example:storage.kube.sys1.xyz
+        lun: 0
+```
+
+### Network
+
+Every network interface is wrapped into a `interface` struct. The `model`
+attribute and all other guest specific attributes are part of this `interface`
+struct. A wrapping `interface.source` struct can contain different source
+structs. They can be traditional interface sources like `bridge`, `direct` or
+`hostedev`, or KubeVirt specializations like `hostNetwork` or `podNetwork`:
+
+```yaml
+devices:
+  interface:
+    source:
+      bridge:
+        dev: mybridge
+    model: virtio
+  interface:
+    source:
+      podNetwork: {}
+    model: virtio
```

Comment:
`model: virtio` is used here, while for video you use `qxl:` rather than `model: qxl`. Why that difference -- is it because of nested values or for other reason? Could you please clarify the libvirt XML <-> KubeVirt YAML mapping? Wouldn't it be better to use something like

    model:
        type: ...

in all cases to keep 1:1 mapping and to avoid mapping confusions and contingent unneeded creativity?

Reply:
@rmohr ?

Reply:
Wherever it is not too much work to explicitly have sub-structs, we should do that. Also see the VolumeSource example above. It is just feated that for most guest-side devices it would be a tremendous amount of work to untangle this. For video it is very easy to do. If it is too confusing, we can leave it with the type, since kubevirt does not have to add it's own extensions there.